### PR TITLE
Check that includes exist in addition to normal modules

### PR DESCRIPTION
**Pull request by narthur157**
Spent some time tracking down a build error yesterday because I had ractive-partials instead of rap in the include for the common bundle. Got to pre before this failed, so it would be nice for this to fail in singularity.

Was thinking about doing excludes too, would be easy enough, but wasn't completely sure about it.

Also, worth noting that grunt-behance has fixed versioning for this so that would need to be updated for this to do something.

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -108,4 +108,47 @@ describe('buildLayerLint', function() {
...
+    it('throws an error when it does not exist', function() {
+      mockFs();
+
+      expect(function() {
+        buildLayerLint('build.json');
+      }).to.throw(/no such file or directory/);
+    });
+  });
+
+  describe('given a build.json with includes where all includes are not defined', function() {
```
Comment (Attamusc) [inline test/index.js]: Could this be a context in the previous describe?

File: index.js [hunk 2]
```diff
@@ -18,9 +18,17 @@ module.exports = function(configPath, modulesRoot) {
...
+    names.push(module.name);
 
-  return config.modules.reduce(function(undefinedModules, module) {
-    var name = module.name;
+    if (module.include) {
+      names = module.include.concat(names);
+    }
+
+    return names;
+  }, []);
```
Comment (Attamusc) [inline index.js]: This will lead to lots of potential duplicates

File: test/index.js [hunk 3]
```diff
@@ -108,4 +108,47 @@ describe('buildLayerLint', function() {
...
+      });
+
+      expect(buildLayerLint('build.json')).to.deep.equal(['fooMod', 'foo', 'bar', 'common']);
+    });
+  });
+
+  describe('given a build.json with includes where all includes are defined', function() {
+    it('returns an empty array', function() {
+      mockFs({
+        'build.json': fs.readFileSync(path.join(fixturePath, 'simple/build.json')),
```
Comment (mrjoelkemp) [inline test/index.js]: Wrong fie?

Comment (narthur157) [inline test/index.js]: definitely the wrong file

**Comment by mrjoelkemp** [comment]
Adding the `exclude`s to this would be great as well for completeness. We typically copy/paste that exclude block, but it's great to have tooling check this.

**Comment by narthur157** [comment]
Right now this reports the log include in the common bundle as not being found, so this isn't actually working as much as I thought

**Comment by narthur157** [comment]
@mrjoelkemp I'll try adding excludes

File: index.js (hunk 2 above)
Comment (mrjoelkemp) [inline index.js]: Yes, probably reduce it into an object with keys equal to the module/include/exclude name and the key being `1` or `true` (some arbitrary value).

Then `moduleNames = Object.keys(thatObject);`

Comment (Attamusc) [inline index.js]: I like that. If only `Set` was a real boy

**Comment by Attamusc** [comment]
Second the `exclude` addition; should actually be an easier feature to implement as well

File: index.js [hunk 4]
```diff
@@ -17,11 +17,23 @@ module.exports = function(configPath, modulesRoot) {
...
+  var moduleNames = config.modules.reduce(function(names, module) {
+    names.push(module.name);
 
-  return config.modules.reduce(function(undefinedModules, module) {
-    var name = module.name;
-    var resolvedName = path.join(modulesRoot, resolveModuleName(config, name));
+    if (module.include) {
+      names = arrayUnion(names, module.include);
+    }
+    if (module.exclude) {
```
Comment (Attamusc) [inline index.js]: Wouldn't an exclude remove it from the group?

Comment (narthur157) [inline index.js]: names is the array of module names to check for the existence of. The idea being, if a module doesn't exist, you shouldn't be excluding it in the first place. I'd guess whether or not that should be the case is open to interpretation, but if it is the case we don't need to check for excludes existing in the first place

**Comment by mrjoelkemp** [comment]
:+1:

**Comment by narthur157** [comment]
This should be 0.2 when it's released because of ^ (grunt-behance pr)

**Comment by Attamusc** [comment]
:+1:

**Pull request merged**